### PR TITLE
Add StepwiseTuner and StepwiseLightGBMTuner

### PR DESCRIPTION
**Pull request by jeffzi**
## Motivation

This PR is related to the discussion in #1632. The steps of `LightGBMTuner` are fixed and its implementation
is tightly coupled to a sequence of pre-defined steps, and to LightGBM in general.

The objective of this PR is twofold:

1. Define a customizable stepwise approach that could be used for multiple integration targets.
2. Refactor the LightGBM tuners to conform to 1. and test that its design is relevant.

### 1. StepwiseTuner

Introduce new concepts:

- `Step`: A [Step](https://github.com/jeffzi/optuna/blob/b96f4077b9958e308fea9e88eb47cb68688145b6/optuna/stepwise.py#L112) encampsulates the specific parameters of one specific optimization task.
- `StepwiseTuner`: A [StepwiseTuner](https://github.com/jeffzi/optuna/blob/b96f4077b9958e308fea9e88eb47cb68688145b6/optuna/stepwise.py#L238) corresponds to a sequence of optimization tasks (steps).
- `Step Objective`: Similar to a regular study objective but takes an extra dictionary of parameters to optimize.

A step holds:

- sampler
- pruner
- n_trials
- timeout
- distributions: A dictionary whose key and value are a parameter name and the corresponding distributions (`BaseDistribution`) to suggest from.

`BaseDistribution` is not usually exposed to end-users elsewhere in Optuna but it allows to defer the "suggestion" to a later time.
Implementations should provide reasonable pre-defined steps.

I chose to replicate the `Study` interface to feel natural. `StepwiseTuner` has a method `optimize()` that accepts the same 
arguments as `Study.optimize`, except the objective `func`. The `timeout` and `n_trials` passed to `optimize` controls the global budgets for all steps. 

`StepwiseTuner` also adds a pre-step that runs a single trial with default parameters to establish a baseline. Some users reported that LightGBMTuner does not always improve upon defaults, this solves the issue.

### 2. StepwiseLightGBMTuner & StepwiseLightGBMTuner

They subclass `StepwiseTuner`. [Pre-defined steps](https://github.com/jeffzi/optuna/blob/stepwise/optuna/integration/_lightgbm_tuner/steps.py) are offered, with the [default sequence](https://github.com/jeffzi/optuna/blob/d7b756371e453aeb5797b17770ba5d55b2f7e48e/optuna/integration/_lightgbm_tuner/steps.py#L186) replicating `LightGBMTuner`'s steps.

In addition to customizable steps, this implementation is further streamlined thanks to:
- Inference of the study's direction from a LightGBM [callback](https://github.com/jeffzi/optuna/blob/b96f4077b9958e308fea9e88eb47cb68688145b6/optuna/integration/_lightgbm_tuner/stepwise.py#L87). It decouples the tuner greatly from LightGBM internals and makes the code more robust.
- Serialization of the model as a string via LightGBM's [model_to_string](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.Booster.html#lightgbm.Booster.model_to_string). Consequently, I've not implemented file serialization. This is probably up to debate.

`StepwiseLightGBMTuner` fixes:
- #1351, #1636: `feval` is verified to work in the tests.
- #1550: parametrizable n_trials per step
- #1858: MAPE metric should be minimized, not maximized.
- Accepts a list of `metric` params but warns that only the first one will be used for optimization.

Lastly, I did not implement `sample_size`. From the code, I understand that it takes the last 'sample_size' rows of the dataset.
In my opinion, this could be easily handled by the user, not the tuner.

Example:
```python
import lightgbm as lgb
import numpy as np
import sklearn.datasets
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

from optuna.integration.lightgbm import StepwiseLightGBMTuner, step_feature_fraction


data, target = sklearn.datasets.load_breast_cancer(return_X_y=True)
train_x, val_x, train_y, val_y = train_test_split(data, target, test_size=0.25, random_state=42)
dtrain = lgb.Dataset(train_x, label=train_y)
dval = lgb.Dataset(val_x, label=val_y)

params = {
    "objective": "binary",
    "metric": "binary_logloss",
    "boosting_type": "gbdt",
    "verbosity": -1,
}
tuner = StepwiseLightGBMTuner(
    params,
    dtrain,
    steps=[step_feature_fraction(n_trials=2)], # use a predefined step
    valid_sets=[dval],
    verbose_eval=False,
    num_boost_round=50,
    early_stopping_rounds=5,
)
tuner.optimize()

model = tuner.get_best_booster()
prediction = np.rint(model.predict(val_x, num_iteration=model.best_iteration))
accuracy = accuracy_score(val_y, prediction)

print("Best params:", tuner.best_params)
print("  Accuracy = {}".format(accuracy))
```
### Other considerations

I wrote tests and docstrings but did not add the documentation to sphinx (waiting for your opinion on the PR).

I realize it may not be ideal to break the `LightGBMTuner` API but I think the proposed API homogenizes better with the rest of Optuna. If the PR is accepted, I propose to start by keeping both `LightGBMTuner` and `StepwiseLightGBMTuner` and scheduling `LightGBMTuner`'s deprecation.

The interface to scikit-learn and the alias for `lgb.train` should also be refactored in the future.

That's all for my proposal. I hope that you like it and that it opens a good discussion. :tada:

### CODE REVIEW

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your great PR.
It contains not only a customizable version of `LightGBM`, but also a general interface which enables step-wise optimization for any objective functions.

I think it takes time to check all the lines, so let me leave just an initial comment.

To understand the usage of `StepwiseLightGBMTuner`, I created an example based on your example in this PR description and `examples/lightgbm_tuner_simple.py`. 
The main change from your example is to change the argument of `StepwiseLightGBMTuner.__init__`. More concretely, I removed `steps` and `num_boost_rounds`, and I changed the values of `verbose_eval` and `early_stopping_rounds`.

```python
    model = lgb.StepwiseLightGBMTuner(
        params, dtrain, valid_sets=[dtrain, dval], verbose_eval=100, early_stopping_rounds=100
    )

    model.optimize()
```

Do you think the optimization steps (i.e., sampler, search space, and `n_trials`) are identical to the [original example](https://github.com/optuna/optuna/blob/master/examples/lightgbm_tuner_simple.py)?

**Comment by jeffzi** [comment]
If you don't provide `steps`, `StepwiseLightGBMTuner` will fall back to the [default steps](https://github.com/jeffzi/optuna/blob/d7b756371e453aeb5797b17770ba5d55b2f7e48e/optuna/integration/_lightgbm_tuner/steps.py#L186) which replicate the original `LightGBMTuner`, including `n_trials`.

In my own tests, the results are close but not identical because:
* `StepwiseTuner` runs one `Trial` with default hyperparameters to avoid performance regression compared to non-optimized hyperparameters. See code [here](https://github.com/jeffzi/optuna/blob/d7b756371e453aeb5797b17770ba5d55b2f7e48e/optuna/stepwise.py#L366).
* None of the approaches set the samplers'seeds. I believe this is something `StepwiseTuner` could potentially do. The problems are that 1) samplers don't have a common method `set_seed` and 2) it would not work if `n_jobs` > 1.

**Comment by toshihikoyanase** [comment]
Thank you for your clear and concise answer.

> `StepwiseLightGBMTuner` will fall back to the default steps which replicate the original `LightGBMTuner`, including n_trials, sampler and search space.
> - `StepwiseTuner` runs one `Trial` with default hyperparameters to avoid performance regression compared to non-optimized hyperparameters.

Cool. I feel the difference between `LightGBMTuner` and `StepwiseLightGBMTuner` is small, and the change to evaluate the baseline configuration is reasonable. As suggested in the description, I also feel `StepwiseLightGBMTuner` can replace the current `LightGBMTuner`. I'd like to check the feasibility of the plan.

> - None of the approaches set the samplers'seeds.

Good point! Users sometimes make optimization reproducible, and I think the proposal is reasonable. I also check if this suggestion can be supported in the new API.

**Comment by jeffzi** [comment]
I found another difference with `LightGBMTuner`. 

`LightGBMTuner` always picks the last `valid_sets` for the optimization (#1984) whereas `StepwiseLightGBMTuner` always picked the first `valid_sets`. 

The reasoning for picking the first `valid_sets` is to follow `first_metric_only`, which is required by `StepwiseLightGBMTuner`. I found out that LightGBM automatically ignores training data in valid_sets. I've now implemented this behavior for `StepwiseLightGBMTuner`. It was a bit tricky because `lgb.train` identifies training sets in `valid_sets` with the `is` operator. I had [to make sure to keep the same reference](https://github.com/jeffzi/optuna/blob/9c064b53ad68904abda203122ed45b1c73261718/optuna/integration/_lightgbm_tuner/stepwise.py#L88) for all train sets when `valid_sets` is copied.

If multiple `valid_sets` are detected, a warning explains that the first non-training set will be used.
